test/token/ERC20WithPermit.test.js

```diff
@@ -0,0 +1,1102 @@
+    expect(await token.symbol()).to.equal("MT")
+  })
+
+  it("should have 18 decimals", async () => {
+    expect(await token.decimals()).to.equal(18)
+  })
+
+  describe("totalSupply", () => {
+    it("should return the total amount of tokens", async () => {
+      expect(await token.totalSupply()).to.equal(initialSupply)
+    })
+  })
+
+  describe("permit_typehash", () => {
+    it("should be keccak256 of EIP2612 Permit message", async () => {
+      const expected = ethers.utils.keccak256(
+        ethers.utils.toUtf8Bytes(
+          "Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)"
+        )
+      )
+      expect(await token.PERMIT_TYPEHASH()).to.equal(expected)
+    })
+  })
+
+  describe("domain_separator", () => {
+    it("should be keccak256 of EIP712 domain struct", async () => {
+      const keccak256 = ethers.utils.keccak256
+      const defaultAbiCoder = ethers.utils.defaultAbiCoder
+      const toUtf8Bytes = ethers.utils.toUtf8Bytes
+
+      const expected = keccak256(
+        defaultAbiCoder.encode(
+          ["bytes32", "bytes32", "bytes32", "uint256", "address"],
+          [
+            keccak256(
+              toUtf8Bytes(
+                "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
+              )
+            ),
+            keccak256(toUtf8Bytes("My Token")),
+            keccak256(toUtf8Bytes("1")),
+            hardhatNetworkId,
+            token.address,
+          ]
+        )
+      )
+      expect(await token.DOMAIN_SEPARATOR()).to.equal(expected)
+    })
+  })
+
+  describe("balanceOf", () => {
+    context("when the requested account has no tokens", () => {
+      it("should return zero", async () => {
+        expect(await token.balanceOf(anotherAccount.address)).to.equal(0)
+      })
+    })
+
+    context("when the requested account has some tokens", () => {
+      it("should return the total amount of tokens", async () => {
+        expect(await token.balanceOf(initialHolder.address)).to.equal(
+          initialSupply
+        )
+      })
+    })
+  })
+
+  describe("transfer", () => {
+    context("when the recipient is not the zero address", () => {
+      context("when the sender does not have enough balance", () => {
+        const amount = initialSupply.add(1)
+
+        it("should revert", async () => {
+          await expect(
+            token.connect(initialHolder).transfer(recipient.address, amount)
+          ).to.be.revertedWith("Transfer amount exceeds balance")
+        })
+      })
+
+      context("when the sender transfers all balance", () => {
+        const amount = initialSupply
+
+        it("should transfer the requested amount", async () => {
+          await token.connect(initialHolder).transfer(recipient.address, amount)
+
+          expect(await token.balanceOf(initialHolder.address)).to.equal(0)
+
+          expect(await token.balanceOf(recipient.address)).to.equal(amount)
```

Comment:
Can we also validate that the `totalSupply` remained unchanged? This could be included in an `afterEach` block for all `transfer` tests.

```js
    afterEach(async () => {
      expect(await token.totalSupply()).to.equal(initialSupply)
    })
```